go/vt/vttablet/tabletserver/system_health.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tabletserver
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/shirou/gopsutil/cpu"
+
+	"vitess.io/vitess/go/sync2"
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/tabletenv"
+)
+
+const (
+	systemHealthMonitorCPUSampleWindow = time.Millisecond * 200
+	systemHealthMonitorInterval        = time.Millisecond * 1000
+)
+
+// systemHealthCollector is a collector for system health.
+type systemHealthCollector struct {
+	config          *tabletenv.TabletConfig
+	cpuSampleWindow time.Duration
+	cpuUsage        sync2.AtomicFloat64
```

Comment:
I think we should move away from using `sync2` now that Go has built in easier to use atomics. We also should switch all existing `sync2` usage but that's a separate change then.

Reply:
@dbussink ahh ok, I thought maybe `sync2` was doing something special 😄

Would you prefer plain `sync/atomic` to be used here?

Reply:
`sync2` is something that has existed since before Go 1.19 came out which added much nicer usable atomics compared to what the Go provided before. See https://golang.org/doc/go1.19#atomic_types

So we haven't done the work yet to migrate, but I've opened #12391 to see what the fallout of that is going to be.

Reply:
Sounds good 👍. One thing it lacks that `sync2` does well is support for `float64` values 🤔

Reply:
> Sounds good 👍. One thing it lacks that `sync2` does well is support for `float64` values 🤔

I wouldn't really classify it as `well` though, since there's a reason why Go doesn't have that, see https://github.com/golang/go/issues/21996.

But you can use the  same trick as for the single other float64 case I found in https://github.com/vitessio/vitess/pull/12391.